rfcs/20200826-keras-nlp-scoping-design.md

```diff
@@ -0,0 +1,108 @@
+They also raise the issue of API standardization.
+
+To fix this, we want machine learning engineers to have access to a standard Keras-native,
+optimized, and well-tested set of components to build their Transformer-based (and beyond) NLP workflows.
+
+This provides key user benefits:
+
+- The package would be first-party and thus always up to date with modern best practices.
+- High code quality and testing standards and strict quality control: same level of trust as core Keras
+- A shared API standard across the community
+- Ability for the open-source community to build more advanced solutions *on top* of this package instead of reinventing it
+- Ability for research scientists to benefit from subclassing and customizing base components to quickly test new research ideas
+
+## Design Proposal
+
+`keras-nlp` will include most standard Transformer-based modules, specifically:
+
+- Keras layer components such as Transformer encoder and decoder blocks.
+- Keras task components such as masked language, span labeler and named entity recognition.
+- Tensorflow operations such as beam search.
+- Keras optimizer utilities such as learning rate schedules widely used.
+- Data loader and preprocessing for different dataset, such as SQUAD, GLUE.
+
+## Success criteria for keras-nlp
+
+- Reusable and standardized components that cover the above
+- Easy-to-use API
+- Models run on CPU/GPU/TPU seamlessly
+- State of the art performance
+- Models can be readily deployed to production
+
+## Boundaries between keras-nlp and tf.text
+
+- `tf.text` will contain all pre-processing operations, such as WordPiece Tokenizer, n-grams, that handles strings.
+- `keras-nlp` will contain modeling components that cover workflows past the tokenization stage.
+
+## Boundaries between `keras-nlp` and TensorFlow Addons:
+
+- Highly experimental modeling, layers, losses, etc, live in Addons (e.g. newly published research code).
+- Components from Addons will graduate to Model Garden, given they get sufficient usage,
+and given that they work on CPU/GPU/TPU. The API interface will remain experimental for a short time after graduation,
+so as to leave us the option to make changes based on user feedback.
+
+## Boundaries between keras-nlp and Model Garden
+
+- End to end modeling workflow and model specific details live in Model Garden
+- Model garden will re-use most of the building blocks from keras-nlp
+- Components from Model Garden can graduate to keras-nlp, given they get sufficient usage,
+and given that they work on CPU/GPU/TPU. The API interface should remain stable after graduation.
+
+# Boundaries between keras-nlp and core Keras
+
+- `keras-nlp` will contain NLP-specific components
+(e.g. the `MultiHeadAttention` layer may be used outside of NLP, and thus is shipping in core Keras).
+- Components from keras-nlp can graduate to Keras core, given its usage expands beyond
+ natural language processing.
+
+## Dependencies
+
+- Tensorflow version >= 2.4
```

Comment:
Could we elaborate on why 2.4 is the minimum version to utilize this library?

Reply:
We need support of several advanced features to make this very readable: 
1) kwargs as call signature, it is currently available in nightly (for sometime), 2.3 might already supported
2) dict inputs and dict outputs & saved model

2.4 might be the first stable release that these are supported, but we can also claim that statement for >= a certain nightly build